File: C_test/C++/src/3.cpp
#include <cassert>
#include <iostream>
#include <queue>
#include <sstream>
#include <unordered_map>

struct TreeNode {
    char      val;
    TreeNode *left;
    TreeNode *right;

    TreeNode(char x)
        : val(x)
        , left(nullptr)
        , right(nullptr) {}
};

// 递归构建二叉树
TreeNode *buildTree(const std::string &inorder, const std::string &postorder, int in_start, int in_end,
                    int post_start, int post_end, const std::unordered_map<char, int> &inorder_map) {
    if (in_start > in_end || post_start > post_end) {
        return nullptr;
    }

    char      root_val = postorder[post_end];
    TreeNode *root     = new TreeNode(root_val);

    int root_index = inorder_map.at(root_val);
    int left_size  = root_index - in_start;

    root->left  = buildTree(inorder, postorder, in_start, root_index - 1, post_start,
                            post_start + left_size - 1, inorder_map);
    root->right = buildTree(inorder, postorder, root_index + 1, in_end, post_start + left_size, post_end - 1,
                            inorder_map);

    return root;
}

// 层次遍历
void bfs(TreeNode *root) {
    if (!root) {
        return;
    }

    std::queue<TreeNode *> queue;
    queue.push(root);

    while (!queue.empty()) {
        TreeNode *curr = queue.front();
        queue.pop();
        std::cout << curr->val;

        if (curr->left) {
            queue.push(curr->left);
        }
        if (curr->right) {
            queue.push(curr->right);
        }
    }
}

void printLevelOrder(const std::string &inorder, const std::string &postorder) {
    int                           n = inorder.size();
    std::unordered_map<char, int> inorder_map;
    for (int i = 0; i < n; ++i) {
        inorder_map[inorder[i]] = i;
    }

    TreeNode *root = buildTree(inorder, postorder, 0, n - 1, 0, n - 1, inorder_map);

    bfs(root);
}

void testPrintLevelOrder(const std::string &postorder, const std::string &inorder,
                         const std::string &expected_output) {
    std::stringstream input;
    std::stringstream output;

    // 重定向输入输出流
    std::streambuf *orig_input  = std::cin.rdbuf(input.rdbuf());
    std::streambuf *orig_output = std::cout.rdbuf(output.rdbuf());

    // 输入测试用例
    input << postorder << " " << inorder;

    // 执行被测函数
    printLevelOrder(inorder, postorder);

    // 还原输入输出流
    std::cin.rdbuf(orig_input);
    std::cout.rdbuf(orig_output);

    // 使用 assert 断言输出结果是否与预期相符
    assert(output.str() == expected_output);
}

void test() {
    {
        std::string postorder = "HIDJEBFGCA", inorder = "HDIBJEAFCG", expected = "ABCDEFGHIJ";
        testPrintLevelOrder(postorder, inorder, expected);
    }
    std::cout << "test passed\n";
}

int main() {
    test();

    std::string inorder, postorder;
    std::cin >> postorder >> inorder;

    printLevelOrder(inorder, postorder);

    return 0;
}

File: C_test/python/3.py
from collections import deque


# 定义了一个 TreeNode 类，表示二叉树的节点
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val  # 值
        self.left = left  # 左子节点
        self.right = right  # 右子节点


def bfs(root):
    if not root:  # 判断根节点是否为空
        return
    queue = deque([root])  # 创建一个双向队列，并将根节点加入队列中
    while queue:
        level_size = len(queue)  # 记录当前层的节点个数level_size
        temp = []
        for _ in range(level_size):  # 循环level_size次
            current_node = queue.popleft()  # 从队列中取出当前节点current_node
            temp.append(current_node.val)
            if current_node.left:  # 如果当前节点存在左子节点，则将左子节点加入队列中
                queue.append(current_node.left)
            if current_node.right:  # 如果当前节点存在右子节点，则将右子节点加入队列中
                queue.append(current_node.right)
        for i in temp:  # 循环结束后，将temp列表中的节点值依次打印出来。循环继续直到队列为空
            print(i, end="")


# 用于构建二叉树
def buildTree(inorder, postorder):  # 该函数接受两个参数inorder和postorder，分别表示中序遍历和后序遍历的结果
    if not inorder and not postorder:
        return
    root = TreeNode(postorder[-1])  # 根节点root，为后序遍历结果的最后一个元素
    idx = inorder.index(postorder[-1])  # 找到根节点在中序遍历结果中的索引idx

    # 递归调用buildTree函数构建左子树和右子树
    root.left = buildTree(inorder[:idx], postorder[:idx])
    root.right = buildTree(inorder[idx + 1 :], postorder[idx : len(postorder) - 1])
    return root


a, b = input().split()
postorder = list(a)
inorder = list(b)

root = buildTree(inorder, postorder)

bfs(root)
